### ISSUE TITLE
accessing tfr data

### ISSUE THREAD

**Issue by mattansb** [craddm/eegUtils#61]
Hi Matt,

This is a really stellar package! Can't wait to see how far it goes (please help me and others never use Matlab again!)

I've been playing around with the functions, and I can't seem to access time-frequency data:

``` r
library(eegUtils)
#> Make sure to check for the latest development version at https://github.com/craddm/eegUtils!
#> 
#> Attaching package: 'eegUtils'
#> The following object is masked from 'package:stats':
#> 
#>     filter

tfr <- compute_tfr(
  data = eegUtils::demo_epochs,  
  foi = c(4, 30),
  n_freq = 10,
  n_cycles = 10
)
#> Output frequencies: 4 6.89 9.78 12.67 15.56 18.44 21.33 24.22 27.11 30
#> Removing channel means per epoch...
#> Warning: The `x` argument of `as_tibble.matrix()` must have column names if `.name_repair` is omitted as of tibble 2.0.0.
#> Using compatibility `.name_repair`.
#> This warning is displayed once every 8 hours.
#> Call `lifecycle::last_warnings()` to see where this warning was generated.

tfr$signals[1:10,1:2,1:2]
#> , , frequency = 4
#> 
#>               electrode
#> time           A5 A13
#>   -0.197265625 NA  NA
#>   -0.189453125 NA  NA
#>   -0.181640625 NA  NA
#>   -0.173828125 NA  NA
#>   -0.166015625 NA  NA
#>   -0.158203125 NA  NA
#>   -0.150390625 NA  NA
#>   -0.142578125 NA  NA
#>   -0.134765625 NA  NA
#>   -0.126953125 NA  NA
#> 
#> , , frequency = 6.88888888888889
#> 
#>               electrode
#> time           A5 A13
#>   -0.197265625 NA  NA
#>   -0.189453125 NA  NA
#>   -0.181640625 NA  NA
#>   -0.173828125 NA  NA
#>   -0.166015625 NA  NA
#>   -0.158203125 NA  NA
#>   -0.150390625 NA  NA
#>   -0.142578125 NA  NA
#>   -0.134765625 NA  NA
#>   -0.126953125 NA  NA
```

But it still seems to be plotting (something?):

```R
plot_tfr(tfr)
```

![](https://i.imgur.com/UewlBud.png)

<sup>Created on 2020-05-12 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by craddm** [comment]
Hi Mattan,

Thanks :)

The output you're seeing is sensible behaviour, which I should document better :) The NAs are because the data at that point is essentially unresolvable with the settings you have. When you specify that the wavelets should be 10 cycles per frequency long, think about the length of data that requires: 10 cycles at 10 Hz is 1 second. The epochs in `demo_epochs` are only ~650 ms long. The number of cycles is constant at each frequency, so the window scales in time according to frequency. 10 cycles at 20 Hz is 500 ms, so up at 20 Hz you can resolve quite a few timepoints even with ~650 ms epochs. If you drop the n_cycles - 3 or 4 is generally fine for frequencies < 30 Hz tbh, but wavelets are often used for gamma band where 7 cycles is pretty common - 

Basically, the function sets the output to NA whenever the window would extend beyond the temporal boundaries of the data. In principle, you could return values for those timepoints, as you can just zero-pad or whatever - but you get boundary effects etc, so it makes for unreliable estimates at those time points.

There's a point in the [Fieldtrip](http://www.fieldtriptoolbox.org/tutorial/timefrequencyanalysis/) tutorial on TFR that mentions these boundary effects too, and you can see the same behaviour there.

Cheers,
Matt

**Comment by mattansb** [comment]
Ah got it, so I'm just looking at the "edges" (that are NA) of the tfr data set.

Thanks!

Awesome package!
